### DISCUSSION TITLE
Starting from private and public pem files

### DISCUSSION THREAD

**Discussion by FrankReh** [dalek-cryptography/curve25519-dalek#577] [Q&A] [answered]
I was trying to start from private and public pem files of EC key pairs, with the hope of building a client/server model similar to WireGuard's, where the pem files are used as the starting point of a short-lived secret key exchange but I thought I should be able to use a pure Rust crate and not have to resort to using the Rust openssl crate for reading the two types of pem file. The pem crate decodes the files into tag and contents nicely, but then I became stuck.

Is the recommendation to use the openssl after all to convert to private and public keys, or is there a method here or another Rust crate that would be suitable?

**Comment by tarcieri** [comment, marked as answer]
`ed25519-dalek` supports this via the `pkcs8` feature.

There's no similar feature in e.g. `x25519-dalek` however it wouldn't be too hard to add using the `pkcs8` crate.

**Reply by FrankReh** [reply]
Thanks for the quick answer. After looking at some of ed25519-dalek, I realized I was trying to solve the wrong problem. WireGuard doesn't deal with pem files or PKCS#8 encodings. It encodes the 32 bytes of private or public key as base64 and that's that. I'm happy to do the same.
